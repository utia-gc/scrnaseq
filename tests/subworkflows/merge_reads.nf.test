nextflow_workflow {

    name "Test Workflow Merge_Reads"
    script "subworkflows/merge_reads.nf"
    workflow "Merge_Reads"

    test("Merge_Reads merges read channels for single end reads when set to not cat reads") {

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [
                            sampleName:   "SRR1066657",
                            sampleNumber: "3",
                            lane:         "001",
                            readType:     "single"
                        ],
                        file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR1066657_S3_L001_R1_001.fastq.gz")
                    ],
                    [
                        [
                            sampleName:   "SRR1066657",
                            sampleNumber: "3",
                            lane:         "002",
                            readType:     "single"
                        ],
                        file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR1066657_S3_L002_R1_001.fastq.gz")
                    ]
                )
                input[1] = false
                """
            }
        }

        then {
            assert workflow.success

            with(workflow.out.reads_merged.get(0)) {
                assert size() == 2
                assert get(1).get(0) ==~ /^.*\/SRR1066657_S3_L001_R1_001.fastq.gz/
                assert get(1).get(1) ==~ /^.*\/SRR1066657_S3_L002_R1_001.fastq.gz/
            }

            assert snapshot(workflow.out).match()
        }

    }

    test("Merge_Reads merges and concatenates fastq files for single end reads when set to cat reads") {

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [
                            sampleName:   "SRR1066657",
                            sampleNumber: "3",
                            lane:         "001",
                            readType:     "single"
                        ],
                        file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR1066657_S3_L001_R1_001.fastq.gz")
                    ],
                    [
                        [
                            sampleName:   "SRR1066657",
                            sampleNumber: "3",
                            lane:         "002",
                            readType:     "single"
                        ],
                        file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR1066657_S3_L002_R1_001.fastq.gz")
                    ]
                )
                input[1] = true
                """
            }
        }

        then {
            assert workflow.success

            with(workflow.out.reads_merged.get(0)) {
                assert size() == 2
                assert get(1) ==~ /^.*\/SRR1066657_R1.fastq.gz/
                assert path(get(1)).linesGzip.size() == 400000
            }

            assert snapshot(workflow.out).match()
        }

    }

    test("Merge_Reads merges read channels for paired end reads when set to not cat reads") {

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [
                            sampleName:   "SRR6924569",
                            sampleNumber: "1",
                            lane:         "001",
                            readType:     "paired"
                        ],
                        [
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L001_R1_001.fastq.gz"),
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L001_R2_001.fastq.gz")
                        ]
                    ],
                    [
                        [
                            sampleName:   "SRR6924569",
                            sampleNumber: "1",
                            lane:         "002",
                            readType:     "paired"
                        ],
                        [
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L002_R1_001.fastq.gz"),
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L002_R2_001.fastq.gz")
                        ]
                    ]
                )
                input[1] = false
                """
            }
        }

        then {
            assert workflow.success

            with(workflow.out.reads_merged.get(0)) {
                assert size() == 3
                assert get(1).get(0) ==~ /^.*\/SRR6924569_S1_L001_R1_001.fastq.gz/
                assert get(1).get(1) ==~ /^.*\/SRR6924569_S1_L002_R1_001.fastq.gz/
                assert get(2).get(0) ==~ /^.*\/SRR6924569_S1_L001_R2_001.fastq.gz/
                assert get(2).get(1) ==~ /^.*\/SRR6924569_S1_L002_R2_001.fastq.gz/
            }

            assert snapshot(workflow.out).match()
        }

    }

    test("Merge_Reads merges read channels for paired end reads when set to cat reads") {

        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [
                            sampleName:   "SRR6924569",
                            sampleNumber: "1",
                            lane:         "001",
                            readType:     "paired"
                        ],
                        [
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L001_R1_001.fastq.gz"),
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L001_R2_001.fastq.gz")
                        ]
                    ],
                    [
                        [
                            sampleName:   "SRR6924569",
                            sampleNumber: "1",
                            lane:         "002",
                            readType:     "paired"
                        ],
                        [
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L002_R1_001.fastq.gz"),
                            file("https://github.com/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L002_R2_001.fastq.gz")
                        ]
                    ]
                )
                input[1] = true
                """
            }
        }

        then {
            assert workflow.success

            with(workflow.out.reads_merged.get(0)) {
                assert size() == 2
                assert get(1).get(0) ==~ /^.*\/SRR6924569_R1.fastq.gz/
                assert path(get(1).get(0)).linesGzip.size() == 400000
                assert get(1).get(1) ==~ /^.*\/SRR6924569_R2.fastq.gz/
                assert path(get(1).get(1)).linesGzip.size() == 400000
            }

            assert snapshot(workflow.out).match()
        }

    }

}
