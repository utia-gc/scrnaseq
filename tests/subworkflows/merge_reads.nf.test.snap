{
    "Merge_Reads merges read channels for paired end reads when set to cat reads": {
        "content": [
            {
                "0": [
                    [
                        {
                            "sampleName": "SRR6924569",
                            "sampleNumber": "1",
                            "readType": "paired"
                        },
                        [
                            "SRR6924569_R1.fastq.gz:md5,525ad355d8f72020b2ad716e38a231f3",
                            "SRR6924569_R2.fastq.gz:md5,4be7fdf7176be1216c332c49de3656df"
                        ]
                    ]
                ],
                "reads_merged": [
                    [
                        {
                            "sampleName": "SRR6924569",
                            "sampleNumber": "1",
                            "readType": "paired"
                        },
                        [
                            "SRR6924569_R1.fastq.gz:md5,525ad355d8f72020b2ad716e38a231f3",
                            "SRR6924569_R2.fastq.gz:md5,4be7fdf7176be1216c332c49de3656df"
                        ]
                    ]
                ]
            }
        ],
        "timestamp": "2023-08-15T20:19:09+0000"
    },
    "Merge_Reads merges read channels for single end reads when set to not cat reads": {
        "content": [
            {
                "0": [
                    [
                        {
                            "sampleName": "SRR1066657",
                            "sampleNumber": "3",
                            "readType": "single"
                        },
                        [
                            "/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR1066657_S3_L001_R1_001.fastq.gz",
                            "/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR1066657_S3_L002_R1_001.fastq.gz"
                        ]
                    ]
                ],
                "reads_merged": [
                    [
                        {
                            "sampleName": "SRR1066657",
                            "sampleNumber": "3",
                            "readType": "single"
                        },
                        [
                            "/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR1066657_S3_L001_R1_001.fastq.gz",
                            "/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR1066657_S3_L002_R1_001.fastq.gz"
                        ]
                    ]
                ]
            }
        ],
        "timestamp": "2023-08-15T20:19:09+0000"
    },
    "Merge_Reads merges and concatenates fastq files for single end reads when set to cat reads": {
        "content": [
            {
                "0": [
                    [
                        {
                            "sampleName": "SRR1066657",
                            "sampleNumber": "3",
                            "readType": "single"
                        },
                        "SRR1066657_R1.fastq.gz:md5,ccabf5ada78353bc8c0584940ab92d35"
                    ]
                ],
                "reads_merged": [
                    [
                        {
                            "sampleName": "SRR1066657",
                            "sampleNumber": "3",
                            "readType": "single"
                        },
                        "SRR1066657_R1.fastq.gz:md5,ccabf5ada78353bc8c0584940ab92d35"
                    ]
                ]
            }
        ],
        "timestamp": "2023-08-15T20:19:09+0000"
    },
    "Merge_Reads merges read channels for paired end reads when set to not cat reads": {
        "content": [
            {
                "0": [
                    [
                        {
                            "sampleName": "SRR6924569",
                            "sampleNumber": "1",
                            "readType": "paired"
                        },
                        [
                            "/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L001_R1_001.fastq.gz",
                            "/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L002_R1_001.fastq.gz"
                        ],
                        [
                            "/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L001_R2_001.fastq.gz",
                            "/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L002_R2_001.fastq.gz"
                        ]
                    ]
                ],
                "reads_merged": [
                    [
                        {
                            "sampleName": "SRR6924569",
                            "sampleNumber": "1",
                            "readType": "paired"
                        },
                        [
                            "/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L001_R1_001.fastq.gz",
                            "/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L002_R1_001.fastq.gz"
                        ],
                        [
                            "/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L001_R2_001.fastq.gz",
                            "/utia-gc/ngs-test/raw/ngs/data/reads/raw/SRR6924569_S1_L002_R2_001.fastq.gz"
                        ]
                    ]
                ]
            }
        ],
        "timestamp": "2023-08-15T20:19:09+0000"
    }
}